rust/src/array.rs

```diff
@@ -15,297 +15,421 @@
 // specific language governing permissions and limitations
 // under the License.
 
+///! Array types
+use std::any::Any;
 use std::convert::From;
-use std::iter::Iterator;
+use std::ops::Add;
 use std::rc::Rc;
 use std::str;
 use std::string::String;
 
 use super::bitmap::Bitmap;
-use super::buffer::Buffer;
-use super::list::List;
-
-pub enum ArrayData {
-    Boolean(Buffer<bool>),
-    Float32(Buffer<f32>),
-    Float64(Buffer<f64>),
-    Int8(Buffer<i8>),
-    Int16(Buffer<i16>),
-    Int32(Buffer<i32>),
-    Int64(Buffer<i64>),
-    UInt8(Buffer<u8>),
-    UInt16(Buffer<u16>),
-    UInt32(Buffer<u32>),
-    UInt64(Buffer<u64>),
-    Utf8(List<u8>),
-    Struct(Vec<Rc<Array>>),
-}
+use super::buffer::*;
+use super::builder::*;
+use super::datatypes::*;
+use super::list::*;
+use super::list_builder::*;
 
-macro_rules! arraydata_from_primitive {
-    ($DT:ty, $AT:ident) => {
-        impl From<Vec<$DT>> for ArrayData {
-            fn from(v: Vec<$DT>) -> Self {
-                ArrayData::$AT(Buffer::from(v))
-            }
-        }
-        impl From<Buffer<$DT>> for ArrayData {
-            fn from(v: Buffer<$DT>) -> Self {
-                ArrayData::$AT(v)
-            }
-        }
-    };
+/// Trait for dealing with different types of Array at runtime when the type of the
+/// array is not known in advance
+pub trait Array {
```

Comment:
do you plan to add more methods? such as `slice`, `is_null`, etc? If so, perhaps add a TODO?

Reply:
I don't think we should add methods as we need, not because we think that we may --- under some circumstances --- need them someday. So a `TODO` is not required.